### ISSUE TITLE
no tok events generated - was "Interpreter fails to handle synchronised parallel correctly"

### ISSUE THREAD

**Issue by joey-coleman** [symphonytool/symphony#121]
> The following bug was [originally reported](https://sourceforge.net/p/compassresearch/bugs/121/) on Sourceforge by [uschulze](https://sourceforge.net/u/uschulze/profile/), 2013-10-10 14:46:32.497000:

New Description:

I changed the description of the problem report, because the problem with the parallel was originated in my process specification and not a problem of the interpreter.

The fixed CML specification will be added to the report. See changes in OUTPUT_CTRL, understand what the problem was.

I did not close this report, but changed it, because the other thing that seems to be worth investigating is, that the interpreter does not generate tok events for the faulty specification. This seems to be wrong and not originated by my specification bug.

---

The original description was:

In the specification attached, the main action of the process definition TurnIndCtrlProcess is a composition of several actions. At the top, there is the combination of

((FLASH_CTRL ||| VOLTAGE ) [| {| newFlashState, newVoltage, newLvrState, dummy |} |] OUTPUT_CTRL)

Notice that OUTPUT_CTLR is synchronised with the interleaving of FLASH_CTRL and VOLTAGE on the events of the channels newFlashState, newVoltage, newLvrState and dummy.

When interpreting the process, only a limited number of events can be selected before the process deadlocks. The last event that can be selected normally is dummy and this shows the situation where the interpreter fails to provide the correct set of events for the next step.

I will try to come up with a smaller example of this bug later.

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/121/#ca16) by [uschulze](https://sourceforge.net/u/uschulze/profile/), 2013-10-12 11:19:56.751000:
- **summary**: Interpreter fails to handle synchronised parallel correctly --> no tok events generated - was "Interpreter fails to handle synchronised parallel correctly"
- Description has changed:

Diff:

```

--- old
+++ new
@@ -1,3 +1,14 @@
+New Description:
+
+I changed the description of the problem report, because the problem with the parallel was originated in my process specification and not a problem of the interpreter.
+
+The fixed CML specification will be added to the report. See changes in OUTPUT_CTRL, understand what the problem was.
+
+I did not close this report, but changed it, because the other thing that seems to be worth investigating is, that the interpreter does not generate tok events for the faulty specification. This seems to be wrong and not originated by my specification bug.
+
+----
+The original description was:
+
 In the specification attached, the main action of the process definition TurnIndCtrlProcess is a composition of several actions. At the top, there is the combination of

 ((FLASH_CTRL ||| VOLTAGE ) [| {| newFlashState, newVoltage, newLvrState, dummy |} |] OUTPUT_CTRL)

```

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/121/#4e2d) by [uschulze](https://sourceforge.net/u/uschulze/profile/), 2013-10-12 11:32:31.612000:

The file turnIndCml.fixed.IDLE.cml contains a version of the turn indication case study that is further developed and that does not show the error with the parallel composition any more. It also is generating tok events again. For the first one, I think i understand why, but for the second one I do not.

The relevant changes are of the type that can be seen in the process IDLE (part of OUTPUT_CTRL).

If you are trying to investigate this, I will be happy to help you. Just contact me for a discussion of the problem, whenever you find time addressing it.> Files were attached:
- [turnIndCml.fixed.IDLE.cml](http://sourceforge.net/p/compassresearch/tickets/_discuss/thread/c19da5fc/4e2d/attachment/turnIndCml.fixed.IDLE.cml) (13616 bytes)

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/121/#520c) by [anderskm](https://sourceforge.net/u/anderskm/profile/), 2013-10-14 12:41:40.819000:
- **status**: open --> accepted

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/121/#c3ac) by [anderskm](https://sourceforge.net/u/anderskm/profile/), 2013-10-14 12:58:47.401000:

I have found the cause of this issue, but I'm not totally sure what the correct solution is. 

The problem is in the "TIMER ||| SWITCH_OFF ||| NEW_DIRECTION" part of the model. As far as I understand the semantics of interleaving, then time only passes if all of the interleaved actions allows time to pass. Since the  "NEW_DIRECTION" action at some point evaluates to "Skip" it stops time from passing for the entire interleaving. 

This behavior seems a bit strange to me, so it could be that I misunderstood this part of the semantics, since this means e.g. that "Wait 1 ||| Skip" would  deadlock which seems wrong to me.

I will of course ask the semantics guys about this.

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/121/#9a8e) by [anderskm](https://sourceforge.net/u/anderskm/profile/), 2013-10-15 09:20:27.710000:
- **status**: accepted --> closed

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/121/#c7b4) by [anderskm](https://sourceforge.net/u/anderskm/profile/), 2013-10-15 09:20:28.235000:

This is fixed in 0d3fcb8
